akka-projection-core/src/main/java/akka/projection/internal/spi/ProjectionMetrics.java

```diff
@@ -0,0 +1,49 @@
+package akka.projection.internal.spi;
+
+import akka.actor.ClassicActorSystemProvider;
+
+public interface ProjectionMetrics {
+
+    /**
+     * Must be invoked for each element received from the Source when the Source provides it. If possible,
+     * invoke this method in a stream stage as close to the stream stage where you read from the wire to
+     * measure the parsing, deserializing and other steps of the processing prior to the event handling.
+     *
+     * @param projectionName the projection name
+     * @param projectionKey  the unique key for this Projection instance
+     * @param creationTimestamp if the element traversing the stream contains the creation time, provide it. Set
+     *                          to null if the information is not available.
+     * @param systemProvider a `ClassicActorSystemProvider` for telemetry to extract/set data on the ActorSystem
+     * @return a contextual object. This context must propagate with the elt.
+     */
+    Object onProcessStart(String projectionName,
+                                          String projectionKey,
+                                          Long creationTimestamp,
+                                          ClassicActorSystemProvider systemProvider);
+
+    /**
+     * Must be invoked for each element processed successfully only when the associated offset has been committed.
+     *
+     * @param projectionName the projection name
+     * @param projectionKey  the unique key for this Projection instance
+     * @param context        the contextual object returned by `onProcessStart`
+     * @return a contextual object. The returned instance may not be the received one.
+     */
+    Object onProcessComplete(String projectionName,
+                                             String projectionKey,
+                                             Object context);
+
+    /**
+     * Must be invoked when the stream fails.
+     *
+     * @param projectionName the projection name
+     * @param projectionKey  the unique key for this Projection instance
+     * @param cause the cause of the failure
+     * @param systemProvider a `ClassicActorSystemProvider` for telemetry to extract/set data on the ActorSystem
+     */
+    void onFailure(String projectionName,
+                   String projectionKey,
+                   Throwable cause,
+                   ClassicActorSystemProvider systemProvider);
+
+}
```

Comment:
This first proposal is quite similar to the Lagom `Projections Spi` except for the fact that this is not bound to a type (in Lagom, all instrumented projections have an `Offset`). Because there's no type that's known beforehand, instrumentation produces an `Object` on the first call on the sequence and that instance must be propagated with the `SteamElement`. Because we're only measuring 2 instants of the stream only `onProcessStart` and `onProcessComplete` need to share that `Object` instance.

Then, `onFailure` must be invoked when the stream fails.

This implementation can be invoked like this:

```scala
// based on the akka-projection-slick branch

  private val metrics: ProjectionMetrics = ???

  // Some funtion that may extract the insertionTime from the StreamElement (or return null)
  val insertionTimeExtractor: StreamElement => Long = _ => null
      
  val (killSwitch, stopped) = {
    Source
      .futureSource(futSource)
      // convert to a SourceWithContext
      .asSourceWithContext(elt => 
         metrics.onProcessStart(projectionId, insertionTimeExtractor(elt), systemProvider))
      .mapAsync(1) { elt => processElement(elt) }
      .mapContext(ctx => metrics.onProcessComplete(projectionId, ctx))
      // convert back to a normal Source
      .asSource
      .watchTermination() { (_, f: Future[Done]) =>
        f.recoverWith {
          case t: Throwable =>
            metrics.onFailure(projectionId, t, systemProvider)
            f
        }
      }
      .toMat(Sink.ignore)(Keep.both)
      .run()
  }
```

*NOTE*: the above example can be simplified to invoke `onProcessStart` and `onProcessComplete` immediately before and after the  `processElement` invocation (inside the `mapAsync`) as demoed in the snippet below. I think other projection implementations may have more stages between the `Source` creation and the `processElement` invocation to, for example, parse the `ByteString` from the wire, deserialize the payload, apply event evelutions,... . The purpose of the example above was to use the SPI using `SourceWithContext`.

Next, is a simplified usage of the `onProcessStart`/`onProcessComplete` proposal for the Slick projection PR:

```scala

Source
  .futureSource(futSource)
  .mapAsync(1) { elt =>
    val metricsContext = metrics.onProcessStart(projectionId, insertionTimeExtractor(elt), systemProvider)
    processElement(elt).map { x =>
      metrics.onProcessComplete(projectionId, metricsContext)
      x
    }
  }
  .watchTermination() { (_, f: Future[Done]) =>
    f.recoverWith {
      ... // etc
```

Reply:
How does this work when failures occur?  I see you have an `onFailure` in your contract, but it's unclear when it would be called. Does it need a reference to the object context too?

Reply:
🤦🏼‍♂️
I cut too much out of the snippet. I think something like the following could work:

```scala
.watchTermination() { (_, f: Future[Done]) =>
  f.recoverWith {
    case t: Throwable =>
      metrics.onFailure(projectionId, t, systemProvider)
      f
  }
}
```

Reply:
I see. Should it accept an object context too like `onProcessComplete`?

Reply:
> I see. Should it accept an object context too like `onProcessComplete`?

That is currently not possible. Note that `watchTermination` is not available in `XyzWithContextOps` so, in order to use it we must go back to vanilla `Source`s (dropping the context), unless there's a way to keep it around.

But, is it possible to take the element (or the `(elt,ctx)` tuple in this case) that caused the stream to fail and pass it along with the failure cause?

I don't recall needing a context on the Lagom SPI for telemetry. As long as I passed the `projectionId` and the `ActorSystem` all the necessary information was available to measure the failure (I think).

Reply:
> But, is it possible to take the element (or the `(elt,ctx)` tuple in this case) that caused the stream to fail and pass it along with the failure cause?

Not without user intervention.

> I don't recall needing a context on the Lagom SPI for telemetry. As long as I passed the `projectionId` and the `ActorSystem` all the necessary information was available to measure the failure (I think).

Ok, I was just curious :)